chain/chain/src/chain.rs

```diff
@@ -2278,53 +2276,49 @@ impl Chain {
         tree_nodes: &mut HashMap<(u64, u64), Option<MerkleHash>>,
     ) -> Result<Option<MerkleHash>, Error> {
         if let Some(hash) = tree_nodes.get(&(index, level)) {
-            Ok(hash.clone())
+            Ok(*hash)
+        } else if level == 0 {
+            let maybe_hash = if index >= tree_size {
+                None
+            } else {
+                Some(*self.mut_store().get_block_hash_from_ordinal(index)?)
+            };
+            tree_nodes.insert((index, level), maybe_hash);
+            Ok(maybe_hash)
         } else {
-            if level == 0 {
```

Comment:
Yeah, especially here, I feel this could have been intentionally written with split iffs.